README.md
<p align="center">
<img width="300" src="docs/logo.jpeg">
</p>

# Elsie
Elsie is a framework for **creating slides programmatically** using Python.

**Quick links**
- [**Documentation**](https://spirali.github.io/elsie)
- Demonstration of features ([PDF](examples/bigdemo/bigdemo.pdf), [source code](examples/bigdemo/bigdemo.py))
- [Gallery of Elsie presentations](https://spirali.github.io/elsie/gallery)
- [API reference](https://spirali.github.io/elsie/apidoc)

## Hello world
```python
import elsie

slides = elsie.SlideDeck()

@slides.slide()
def hello(slide):
    slide.text("Hello world!")

slides.render("slides.pdf")
```

## Installation
### Requirements
- Python 3.6+
- Inkscape 1.0+ (required only for Inkscape backend)
    - You can find installation instructions [here](https://wiki.inkscape.org/wiki/index.php/Installing_Inkscape)
    - Versions under 1.0 and above 0.92 might also work, but they are not primarily supported
    - Elsie uses Inkscape in the background, you do not have to know how to use it
- `pdflatex`, `pdf2svg` (required only for `LaTeX` support)

### Installation using pip
```bash
$ pip3 install elsie
```

If you want to use the [`Cairo`](https://spirali.github.io/elsie/userguide/basics/#backends) backend,
install `Elsie` with the `cairo` extra package:
```bash
$ pip3 install elsie[cairo]
```

### Installation using Docker
```bash
$ docker build -t elsie .
```


CHANGELOG.md
Note: This file documents major changes visible to users; see Git history for detailed log.

# v3.4

* Detecting Inkscape on Windows

# v3.3

* Dependencies updated
* Better delimiter settings

# v3.2

## Fixes

* Comptability with Inkscape 1.1

## New features

### New modules in extension added:

* Simple export to pptx (slides as PNG pictures)

```python
from elsie.ext.export import export_pptx

export_pptx(slides, "output.pptx")
```

* Todo placeholder

```python
from elsie.ext.todo import todo_placehodler

todo_placeholder(box)
```

# v3.1

## New features

### Add new experimental Cairo backend
You can use the new Cairo backend to render slides without the need to install `Inkscape`.
To use it, install elsie with the `cairo` extra package:
```bash
$ pip install elsie[cairo]
```

And then pass `CairoBackend` to `SlideDeck`:
```python
import elsie
from elsie.render.backends import CairoBackend

slides = elsie.SlideDeck(backend=CairoBackend())
```

# v3.0

## Breaking changes

### Global instance of Slides is removed and Slides are renamed to SlideDeck

Global slides were removed. You can just create instance of Slides (now SlideDeck)
as follows:

Before:
```python
elsie.slide()
def a_function(slide):
    pass
```

Now:
```python
slides = elsie.SlideDeck()

slides.slide()
def a_function(slide):
    pass
```

### Box arguments are now "keyword only"

Before you could write:

```python
slide.box(50, 0, 100, 100)
```

To avoid confusion, positional arguments are disabled and all arguments have to be named:

```python
slide.box(x=50, y=0, width=100, height=100)
```


## New features

* [New documentation](https://spirali.github.io/elsie/)
* Large speedup in slide rendering by using Inkscape shell interface
* `ext` package with support for [lists](https://spirali.github.io/elsie/userguide/lists/) and [markdown](https://spirali.github.io/elsie/userguide/lists/)
* [Jupyter integration](https://spirali.github.io/elsie/userguide/jupyter/)


# v2.1

## New features

* ORA support
* Method TextStyle.compose()

# v2.0
## Breaking changes
### TextStyles
Text styles are no longer dictionaries. From now on, they will be instances of `elsie.TextStyle`.

New methods for working with styles:

```
box.set_style("new-style", elsie.TextStyle(size=10))
box.update_style("new-style", elsie.TextStyle(bold=True))
```

Methods `new_style` and `derive_style` were removed.

### Line box and Inline box
`line_box` and `inline_box` (previously `text_box`) can be now only used on objects returned by
the `text` or `code` methods.

Before:
```python
b = ...box(...)
b.text(...)
b.line_box(...)
b.text_box(...)
```

Now:
```python
b = ...box(...)
t = b.text(...)
t.line_box(...)
t.inline_box(...)
```

Argument `lines` in `line_box` was renamed to `n_lines`.

## New features
Arguments `above` and `below` in `.box(..)` method. See
[documentation](https://spirali.github.io/elsie/userguide/layout/#modifying-render-order)
for their explanation.
```python
r = b.rect(...)
b.box(..,  below=r)  # The new box is drawn before rectangle
```


makepng.py
import elsie
import subprocess
import datetime

slides = elsie.Slides()


@slides.slide()
def hello(slide):
    this_year = datetime.datetime.now().year
    slide.text(f"Hello from {this_year}!")


@slides.slide()
def boxdemo0(slide):
    slide.box().text("Box 1")
    slide.box().text("Box 2")
    slide.box().text("Box 3")
    slide.rect(color="black", stroke_width=4)


@slides.slide(debug_boxes=True)
def boxdemo1(slide):
    slide.box().text("Box 1")
    slide.box().text("Box 2")
    slide.box().text("Box 3")


@slides.slide(debug_boxes=True)
def sizedemo1(slide):
    slide.box().text("Box 1")
    slide.box(width=300, height=100).text("Box 2")
    slide.box(width="100%").text("Box 3")


@slides.slide(debug_boxes=True)
def composition(slide):
    slide.box(padding=40).box(padding=40).box(padding=40).text("Box 1")


@slides.slide(debug_boxes=True)
def filldemo1(slide):
    slide.box().text("Box 1")
    slide.box(width=300, height=100).text("Box 2")
    slide.box(height="fill").text("Box 3")


@slides.slide(debug_boxes=True)
def padding_demo(slide):
    slide.box(width=200, height=200, p_left=100, name="Top box")
    slide.box(width=200, height="fill", p_y=100, name="Bottom box")


@slides.slide(debug_boxes=True)
def horizontal(slide):
    parent = slide.box(width="100%", height="100%", horizontal=True)
    parent.box().text("Box 1")
    parent.box().text("Box 2")
    parent.box().text("Box 3")


@slides.slide(debug_boxes=True)
def xy(slide):
    b = slide.box(width=100, height=100, name="First")
    slide.box(x=b.x("50%"), y=b.y("50%"), width=100, height=100, name="Second")
    slide.box(x=0, y=b.y("100%"), width="100%", height=200, name="Third")


@slides.slide()
def paiting1(slide):
    slide.box(x="[40%]", y="[40%]", width=300, height=300).rect(bg_color="red")
    slide.box(x="[60%]", y="[50%]", width=300, height=300).rect(bg_color="green")
    slide.box(x="[30%]", y="[60%]", width=300, height=300).rect(bg_color="blue")
    slide.rect(color="black", stroke_width=4)


@slides.slide()
def paiting2(slide):
    red = slide.box(x="[40%]", y="[40%]", width=300, height=300)
    red.rect(bg_color="red")
    slide.box(x="[60%]", y="[50%]", width=300, height=300).rect(bg_color="green")
    slide.box(below=red, x="[30%]", y="[60%]", width=300, height=300).rect(
        bg_color="blue"
    )
    slide.rect(color="black", stroke_width=4)


@slides.slide()
def paiting3(slide):
    slide.box(x="[40%]", y="[40%]", width=300, height=300).rect(bg_color="red")
    slide.box(z_level=1, x="[60%]", y="[50%]", width=300, height=300).rect(
        bg_color="green"
    )
    slide.box(x="[30%]", y="[60%]", width=300, height=300).rect(bg_color="blue")
    slide.rect(color="black", stroke_width=4)


@slides.slide()
def steps(slide):
    slide.box().text("Box 1")
    slide.box(show="2+").text("Box 2")
    slide.box(show="3+").text("Box 3")
    slide.rect(color="black", stroke_width=4)


@slides.slide()
def text1(slide):
    slide.text("Hello world!", elsie.TextStyle(size=70, color="red"))
    slide.rect(color="black", stroke_width=4)


@slides.slide()
def text2(slide):
    default_style = slide.get_style("default")
    default_style.color = "orange"
    slide.set_style("default", default_style)
    slide.text("Hello world!")
    slide.rect(color="black", stroke_width=4)


@slides.slide()
def inline_styles(slide):
    slide.set_style("red", elsie.TextStyle(color="red"))
    slide.text("Normal text ~red{red text} ~tt{Typewriter text}")
    slide.rect(color="black", stroke_width=4)


@slides.slide(debug_boxes=True)
def scale_to_fit(slide):
    slide.box(width=300, height=80).text("Hello world!", scale_to_fit=True)
    slide.box(width=80, height=300).text("Hello world!", scale_to_fit=True)


@slides.slide()
def code(slide):
    slide.box().code(
        "python",
        """
        x = "Elsie"
        print("Hello", x)
    """,
    )
    slide.rect(color="black", stroke_width=4)


@slides.slide()
def textboxes(slide):
    text_item = slide.box().text(
        """This is a long
    text ~#A{that} takes
    3 lines.
    """
    )

    text_item.line_box(2, z_level=-1).rect(bg_color="red")
    text_item.inline_box("#A", z_level=-1).rect(bg_color="green")
    slide.rect(color="black", stroke_width=4)


@slides.slide()
def rectangle(slide):
    box = slide.box(x="[50%]", y="[50%]", width="80%", height=300)
    box.rect(
        bg_color="green",
        color="red",
        stroke_width=10,
        stroke_dasharray="10 4",
        rx=20,
        ry=20,
    )
    slide.rect(color="black", stroke_width=4)


@slides.slide()
def textrect(slide):
    slide.box().rect(bg_color="#aaf").text("Hello!")
    slide.rect(color="black", stroke_width=4)


svgs = slides.render(return_svg=True)  # Creates file 'slides.pdf'
# svgs = slides.render(return_svg=False)  # Creates file 'slides.pdf'

for slide, step, out in svgs:
    name = slide._box.name
    if step == 1:
        filename = "slide_imgs/{}.svg".format(name)
    else:
        filename = "slide_imgs/{}-{}.svg".format(name, step)
    print(name)
    with open(filename, "w") as f:
        f.write(out)
    subprocess.check_call(
        ["inkscape", "-b", "#ffffff", "--export-type", "png", filename]
    )
